Add pydantic schemas settings

kowshikdontu opened the pull request:
issue #19 

- Added Pydantic Schemas
Created app/schemas/otp.py and app/schemas/token.py to define strong request and response models for the OTP API. These schemas enforce input validation, make endpoint contracts clear, and improve the OpenAPI documentation.

- Centralized Configuration with BaseSettings
Introduced app/config.py using BaseSettings (from the pydantic-settings package, as required by Pydantic v2+). This consolidates all configuration, such as Twilio credentials and JWT secrets into a single, type-safe location.

- Twilio Verify v2 Migration
Updated Twilio API calls to use client.verify.v2.services(...) instead of the deprecated v1 version. This aligns with the latest Twilio library and resolves PyCharm deprecation warnings.

- Improved API Documentation
Added response_model and status_code to each endpoint, and provided concise docstrings. This enhances the autogenerated API docs and makes responses and behaviors more predictable for consumers.

- Added __init__.py to All Folders
Included empty __init__.py files in all subfolders (app/, app/routers/, app/schemas/, etc.) to ensure Python treats them as packages. This allows for absolute imports (e.g., from app.schemas.otp import OTPRequest) and resolves module import issues in editors and during deployment.

- improved requirements.txt with versions
  frozen the requirements for  clean installations of pydantic-settings and new twilio versions

<img width="921" height="792" alt="image" src="https://github.com/user-attachments/assets/b2127295-0f3f-4144-8810-4ecb2b7ea2b7" />

<img width="901" height="718" alt="image" src="https://github.com/user-attachments/assets/af4e4226-32c5-4e8c-abeb-18b3f61b6a91" />

Harish-2003 commented:
requriments.txt?
and currently we are planning to integrate with the neon db and will look into it once we integrate it and look into the possibilities of merging the issue

kowshikdontu commented:
ok , after integrating the db, anyhow all the db.py functions also require pydantic schemas in orm_mode for clean validation , i will add them and test the api routes and update the branch sir

if you could assign this schema validation issue to me , I would wait for the db integration too.

thank you for reviewing the pr

kowshikdontu commented:
I made the existing code more enhanced and industry standard sir @Harish-2003 

key improvements:
- created schemas folder with otp_schemas and user_schemas
- added status_codes to routes 
-  changed the query parameter to json body path parameters as phone number is PII and not advisable to use query parameters
- used the get_db func that is idle in database.py as dependency injection in routes and passed it as parameters to functions, which is more standard practice
- added  empty  __init__.py to folder for better modularity
- added access token verification functions and get current user functions in jwt.py for future use

I used postman to verify the routes and created a Twilio service to test all function, screenshots below with a one line description  will helps you to understand the testing.


-   checked  /otp/register with random phone number
<img width="811" height="787" alt="Screenshot 2025-07-27 135202" src="https://github.com/user-attachments/assets/26f9aef3-6fa1-4b73-943a-5f312ee32fc3" />

- checked send-otp function after creating a Twilio service free trail
<img width="813" height="592" alt="Screenshot 2025-07-27 141253" src="https://github.com/user-attachments/assets/093dd5a0-292c-48fc-b6c5-ed7e46396024" />

- checked verify-otp function but with not registered phone number
<img width="825" height="694" alt="Screenshot 2025-07-27 141445" src="https://github.com/user-attachments/assets/1963279d-87db-409e-8608-f4e437ec9783" />

- registered the crct phone number and verified the otp, otp  got verified successfully ( changed the mail as previous mail raised unique constraint violation )
<img width="820" height="814" alt="Screenshot 2025-07-27 141657" src="https://github.com/user-attachments/assets/866d8fd3-4b99-4251-8280-4753ee7ff22b" />

<img width="818" height="829" alt="Screenshot 2025-07-27 141805" src="https://github.com/user-attachments/assets/fae06e50-716f-412a-ad35-e26bbf5bd05f" />

ss of otps received
<img width="540" height="1135" alt="image" src="https://github.com/user-attachments/assets/f735a253-1f23-47b2-9cbf-473b0bf7b656" />

FastAPI server side log
 
<img width="849" height="483" alt="Screenshot 2025-07-27 141953" src="https://github.com/user-attachments/assets/47fb8c9a-2188-46c6-a97e-5b1260431cda" />

used a local .db file for quick testing and the visualization of db table is
<img width="1590" height="586" alt="Screenshot 2025-07-27 142252" src="https://github.com/user-attachments/assets/c7be577c-b2b2-4aea-b986-a901a54a5058" />

from api side it all clean and clear to merge this pull request sir, as it is SQL Alchemy if we just change the database url to neon db , it works the same !   could you please look into this.

kowshikdontu commented:
@Harish-2003
